skipruntime-ts/helpers/src/utils.ts

```diff
@@ -68,3 +70,111 @@ export function reactiveResponseHeader(
 ): [string, string] {
   return ["Skip-Reactive-Response-Token", JSON.stringify(reactiveResponse)];
 }
+
+/*****************************************************************************/
+// Joins
+/*****************************************************************************/
+
+type JoinObject = { value: Json; side: "left" | "right" };
+
+class AddJoinSideField implements Mapper<Json, Json, Json, JoinObject> {
+  constructor(private joinSide: "left" | "right") {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<Json>,
+  ): Iterable<[Json, JoinObject]> {
+    const result: [Json, JoinObject][] = [];
+    for (const v of values) {
+      if (typeof v !== "object") {
+        throw new Error(
+          "joinCollection only works on objects, not: " + JSON.stringify(v),
+        );
+      }
+      result.push([key, { value: v, side: this.joinSide }]);
+    }
+    return result;
+  }
+}
+
+function mergeObjects(object1: JoinObject, object2: JoinObject): Json {
+  const v1 = object1.value;
+  const v2 = object2.value;
+  if (typeof v1 !== "object") {
+    throw new Error(
+      "mergeObjects only works with objects, not: " + JSON.stringify(v1),
+    );
+  }
+  if (typeof v2 !== "object") {
+    throw new Error(
+      "mergeObjects only works with objects, not: " + JSON.stringify(v2),
+    );
+  }
+  const keys1 = Object.keys(v1);
+  if (keys1.some((key) => key in v2)) {
+    throw new Error(
+      "Objects don't have distinct fields: " +
+        JSON.stringify(v1) +
+        " " +
+        JSON.stringify(v2),
+    );
+  }
+  return { ...v1, ...v2 };
+}
+
+class MergeJoinFields implements Mapper<JoinObject, JoinObject, Json, Json> {
+  constructor(private allowMultipleValuesOnBothSides: boolean) {}
+
+  mapEntry(
+    key: Json,
+    values: NonEmptyIterator<JoinObject>,
+  ): Iterable<[Json, Json]> {
+    const result: [Json, Json][] = [];
+    let countLeft = 0;
+    let countRight = 0;
+    for (const value of values) {
+      if (value.side === "left") {
+        countLeft++;
+      }
+      if (value.side === "right") {
+        countRight++;
+      }
+    }
+    if (countLeft > 1 && countRight > 1) {
+      if (!this.allowMultipleValuesOnBothSides) {
+        throw new Error(
+          "More than one value detected on both sides for key: " +
+            JSON.stringify(key),
+        );
+      }
+    }
+    for (const value1 of values) {
+      if (value1.side === "left") {
+        for (const value2 of values) {
+          if (value2.side === "right") {
+            result.push([key, mergeObjects(value1, value2)]);
+          }
+        }
+      }
+    }
+    return result;
+  }
+}
+
+export function joinCollections<
+  K extends Json,
+  V1 extends Json,
+  V2 extends Json,
+>(
+  col1: EagerCollection<K, V1>,
+  col2: EagerCollection<K, V2>,
+  allowMultipleValuesOnBothSides: boolean = false,
```

Comment:
The expectation is that the fields of the values in the left and right value types are disjoint, right? Or is it expected that idiomatic use cases will have non-disjoint types, but only per-key will disjointness will hold?

My understanding is that the simplest case is where both the left and right argument collections map keys to unique values. On the other hand, if the left and right argument collections map keys to potentially multiple values, then the result will map each key to the merge of each pair in the cartesian product of the left values and right values for that key.

Maybe this is just me, but I don't see why the intermediate cases, where only one of the argument collections is multi-valued, are grouped with the simple case rather than the general one. Would it be a simpler interface if there were two booleans, one stating that the left collection is allowed to be multi-valued and the other stating that the right collection is allowed to be multi-valued?

Reply:
The reason is my experience with the SQL engine (which might be flawed). I have seen multiple cases where either side had more than one value, but never something that required a cartesian product that was actually useful. Let's review them, let's assume we have 2 large collections (large in the sense that N*M is too large to be computed).

Example, one / many: A table of users, a table of addresses, a user can have multiple addresses => maximum size of the able = number of address
Example, one / one: A table of users, a table of social security numbers => maximum size of the table = min(users, ssn)
Example, many / many: ??? Usually a contrived example that is only useful to explain the behavior of a join 

I understand the cartesian product in the many/many case is the natural extension of what the join should do. But I have yet to see anything in practice using it. In fact, more often than not, it's a mistake, the query explodes and the user does not know why.

When you think about it, it makes sense. A cartesian product is only going to work for very small sets ...

So I would rather make it explicit. It's not banned. You just have to be explicit about it.